### PR TITLE
feat(Core): Instance saved data implementation

### PR DESCRIPTION
**Pull request by claudiodfc**
## Changes Proposed:
Currently, azerothcore doesn't save gameobject states on instances.
Whenever there's a re-start or crash, the instance's gameobjects and their states aren't saved, producing un-wanted behaviours and blocking instances at times.

Implemented CRUD for new table `instance_saved_go_state_data` that holds the states of gameobjects.

- When worldserver launches and gameobjects are loaded, this will check if this object's state exists on the DB and sets the previous state.
- On instance deletion (reset) these states are also removed based on the instance ID.
- Whenever a gameobject state changes inside a dungeon or raid, we save on the database the set state.

## Issues Addressed:

- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/9422
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/9468
- Closes part of https://github.com/azerothcore/azerothcore-wotlk/issues/9414
- Closes part of https://github.com/azerothcore/azerothcore-wotlk/issues/9529

## SOURCE:

No source

## Tests Performed:

- Tested in-game

## How to Test the Changes:

The following test case should be executed when in a party and solo, to see if the instances are resetting properly.

If possible testing some dungeons like the ones mentioned on the issues would be appreciated.

#### Test case 1
1. .tele culling
2. .go gameobject 67455
3. .gobject activate 67455
4. .ser res 1
5. Re-launch Worldserver
6. Check if book-case is still open
7. Repeat steps 1, 2 and 3 but Reset all instances after repeating the first step.

## Known Issues and TODO List:

- [x] There might be some crash so if possible doing other dungeon test such as the Deadmines would help out. I wasn't able to reproduce a crash again.

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Review by Nyeriah — APPROVED** [review]

**Comment by Ovahlord** [comment]
This whole change is stupid. GameObject states are not suposed to be dumped in database. Door like boss rooms are suposed to be tied to a bosses boss state / encounter. If these doors are not working, it means the boss state handling is messed up or the doors have not been linked.

Outside of that case, regular gameobjects are also not suposed to save their state in DB because instances do reset and anything that is not initialized properly has not been correctly tied to a boss encounter. This all should be handled either by directly linking the gameobjects to it or handle them on spawning, which belongs into the InstanceScript::OnGameObjectCreate hook.

So long story short: this change fixes nothing that can't already be fixed when someone knows their shit and only generates more junk DB data

**Comment by mdX7** [comment]
> Where is this written?

Blizz doesnt write stuff like this, its based on intensive research.

You can handle it like you want, but if you would've done proper research you'd know stuff is tied to boss states.

**Comment by Nyeriah** [comment]
@Ovahlord @mdX7 thanks for your feedback, but toxicity isn't particularly welcome here. Those objects are not necessarily tied to boss states, and honestly, I don't see the issue of doing this differently than TrinityCore does. Instance scripts aren't sniffed.

**Comment by mdX7** [comment]
> I don't see the issue of doing this differently than TrinityCore does. 

You can even do it differently than Blizz as long as the experience of a player is unaffected ingame (i dont mean exp as in required for leveling up). Even going more abstract or reinventing boss states is also an option. 

Personally I don't like saving all changed gob states of certain gob type, it kinda can bloats db unnecessarily and causes undefined behavior for systems utilizing gob states in a different way (boss states e.g.)
Saving only required gobs sounds a bit better to me as in defining them one by one. It may have strange behavior for a bossfight utilizing gobs (maybe lk platform/intermission) when saving every gob statechange and crashing at wrong timing. Keeping on that lk example iirc the outer platform isnt available unless intermission is active, so crashing the server while intermission is active could lead to gob being spawned when server is up again. If it does not do it its properly handled by bossstates and junk might've been written to db since you are not differing between permanent and temporary spawns. Or its indeed skipped due to not satisfying the gob type check (im too lazy to check rn)
Keep in mind this is pure theory and not tested in any way.

Something about codestyle: in cases you do not need a `break` after returning something before. Return stops code execution of the entire function and due to that `break` is unreachable code in this scenario:
```c
case 0:
    return something;
    break; // unreachable/dead code
```

Its not about how tc does stuff, this is a fork you can do w/e you want obviously. *giggles in 400ms delay*

**Comment by Nyeriah** [comment]
Snarky comments don't help, so the question is, and I don't really understand it, if the point isn't helping with anything, why bother going out of your way just to come here and insult others?

**Comment by FrancescoBorzi** [comment]
> This whole change is stupid.

@Ovahlord no one speaks like that at AzerothCore, we don't like any flame attitude or drama. If you want to leave your feedback, you're welcome to do so as long as you respect everyone.

**Comment by mdX7** [comment]
> Snarky comments don't help

Only the very last line of my previous post being sarcastic, but everything before was either feedback or stuff people can discuss we are talking about serverside stuff here afterall. Also initially I only thought about directly boss related doors which obv should be boss state related which you can proof on official servers for **some bosses** by killing, reseting, wiping.

>  if the point isn't helping with anything

Pointing towards problems is actually helping unless you prefer problems appearing after prs were merged.

> why bother going out of your way just to come here

Well my intention was to provide feedback and info to discuss. In a different pr pointing towards a possible crash was also happily ignored lol.

> and insult others?

i'm not really considering any of my posts here an insult but okay, feel free to do so.  
The comment of @claudiodfc which @FrancescoBorzi apparently removed via edit is an insult towards tc however.

But yeah enough offtopic; I'll just leave apparently i'm unwanted here, good luck anyway.

**Comment by Nyeriah** [comment]
I believe there are things that are obvious to you and to your fellow workmates but they might not be as obvious to our contributors, which in majority are not professional developers and don't have vast experience with the game and the engineering behind it. I also believe there are ways of expressing yourself without trying to make less of people around you.

We have contributors of varying skill levels but the majority of them don't have years of emulation experience, we try to help them improve themselves and attempt to point out their mistakes in a didatic way, without demeaning anybody

It's not humanly possible to keep track of everything and sometimes mistakes are made, we try give them a timely response

I believe it's difficult to come to a project you're not familiar with and try to review something, but the reaction you got is none but expected considering how unfriendly some people were (**you** as in the plural *you*).

I sincerly appreciate the helpful remarks, however, sarcasm often overshadows any genuine help attempt. We could certainly do without it and we'd appreciate your help even more that way, definitely.

**Comment by FrancescoBorzi** [comment]
> The comment of @claudiodfc which @FrancescoBorzi apparently removed via edit is an insult towards tc however

@mdX7 which is exactly why I removed it

**Comment by claudiodfc** [comment]
Thanks for the suggestion and review MDX.

**Comment by Lynethris** [comment]
Hi, first review of my test. 

No worldserver, no crash find for the moment. One information for tester, actually you can't use this PR without an existing db  (you got a error with missing tables because you don't have applied the .sql)

Your two cases tests have been done with success. Just one minor issue with deadmines. When  the restart, the door is open but  the object is marked like "locked" as if it's not open. And normally, the door shouldn't be visible because it was destroyed.

After this, I have quickly tested the case where the server crash. It's working too but I've got a strange situation, because when the server restart, players get a rollback x time before the crash. And we can get a situation where: players kill boss 1, they walk and unlock a door. Come a crash, they back x time before and the boss 1 is alive but the door stays open. 

I will continue my tests  in the coming days. (So this PR stay in testing in progress ;) )

**Comment by hanabi5** [comment]
Thank-you for your testing efforts @Lynethris. I can see why the Deadmines issue is happening and I'll try seeking into it. Since it's such a specific issue, might as well deal with it specifically.
Regarding the Boss staying alive after a server crash, that's another issue that I don't plan on covering on this pull-request.

I'm gonna post another test case for Gnomeregan that I'd like you to test as well.
Good work!

**Comment by hanabi5** [comment]
@Lynethris could you please repeat the test case on deadmines?

**Comment by Lynethris** [comment]
> @Lynethris could you please repeat the test case on deadmines?

I'm actually a little busy with the BWL release. I try to confirm you changes this night and I will continue other tests in the end of the week.

**Comment by Lynethris** [comment]
Hi, I just spent a few hours to test your PR. First, the fix for the canon door is good. After: 

- [ ] No crash or worldserver error find
- [ ] Don't break any mechanic
- [ ] When instances reset, the DB is clean
- [ ] Fix the scholomance issue

Minor issue: After a restart/crash, doors stay open but they are selectable/interactive (see image)

You mention that this PR close two issues with gnomeregan and stratholme but I see any changes in their code. Normal ?

After, if I understand only instance that include "gameobject->UpdateSaveToDb(true);" are saved in the DB, right ? It's bind to all the gameobject in the instance or only the gameobject mark by "UpdateSaveToDb" ?

I ask this because I find a strange thing with deadmines. When you enter in the instance 5 entries are saved in the DB (see images). Probably all the doors, but there are 6 doors in the instance. After having opened all the doors and do a restart server, the 2nde, 3rd, and 5rd door stay close (but you can reopen them). The 3rd and 5rd are bind to a lever.

**Comment by hanabi5** [comment]
> You mention that this PR close two issues with gnomeregan and stratholme but I see any changes in their code. Normal ?

I was waiting for those test cases to be validated in order to fix the Stratholme and Gnomeregan doors. Now that it looks good, I'll push some changes for those instances too.

> Minor issue: After a restart/crash, doors stay open but they are selectable/interactive (see image)

I'll have to take change something to prevent this from happening

> After, if I understand only instance that include "gameobject->UpdateSaveToDb(true);" are saved in the DB, right ? It's bind to all the gameobject in the instance or only the gameobject mark by "UpdateSaveToDb" ?

Only the gameobjects that are marked as ```UpdateSaveToDb(true)``` will be saved on the db.

> there are 6 doors in the instance

You're correct on this, I didn't add these doors to the database since you can interact with them, but for immersion sake, I will probably add all doors in order to replicate the saved instance state.

Thank you very much for the tests @Lynethris! I'll give news about Stratholme and Gnomeregan soon. I'll find a way to protect gameobjects from being interactable after their state changes when necessary.

File: src/server/database/Database/Implementation/CharacterDatabase.cpp [hunk 1]
```diff
@@ -599,6 +599,13 @@ void CharacterDatabaseConnection::DoPrepareStatements()
     PrepareStatement(CHAR_SEL_CHAR_SETTINGS, "SELECT source, data FROM character_settings WHERE guid = ?", CONNECTION_ASYNC);
     PrepareStatement(CHAR_REP_CHAR_SETTINGS, "REPLACE INTO character_settings (guid, source, data) VALUES (?, ?, ?)", CONNECTION_ASYNC);
     PrepareStatement(CHAR_DEL_CHAR_SETTINGS, "DELETE FROM character_settings WHERE guid = ?", CONNECTION_ASYNC);
+
+    // Instance saved data. Stores the states of gameobjects in instances to be loaded on server start
+    PrepareStatement(CHAR_SELECT_INSTANCE_SAVED_DATA, "SELECT id, guid, state FROM instance_saved_go_state_data", CONNECTION_SYNCH);
```
Comment (acidmanifesto) [inline src/server/database/Database/Implementation/CharacterDatabase.cpp:604]: I am curious. Why not asynch like the rest?

**Comment by hanabi5** [comment]
> The save seems bind to the lead of the group. So when he leave the group, the DB is cleared. I observed that if the rest of the group interact with a gameobject, a new instance is saved. But after having done a restart in deadmine (after activating the last door), at the relog the door was closed.

The save isn't necessarily bound to the leader of the group. We don't account for players saved on this PR, only for instance IDs. When a player leaves a dungeon group, his instance ID is cleared. So even if other players are inside, they are still in the instance ID that the leader had, only that his ID was cleared due to him abandoning the group which is intended otherwise we would have un-wanted behaviours. Being a leader means nothing here, It's just a coincidence of your test case.

Now, you mentioned the door being closed, but to whom was it closed, the player who left the group or the remaining players on party?

> I think of another problematic situation. It's with raid. Actually, instances no use are clear. But one group can do 3 bosses and decide to comeback one or two days later. This feature could not be used in this case.

I'm not sure I understood what you meant. But let me put in perspective the raids problem.

Currently, If you're doing Molten Core and have no Aqual Quintessences but you did kill all bosses to Majordomo, the Runes get activated for free because the code activates the runes based on bosses killed, allowing players to finish the Raid even if they didn't douse any Runes. This was a change that they had to implement in order to make sure that even after a crash, people could finish Majordomo.
The use for this pull request on raids is being able to save the Rune's states and not providing free encounters to players.

Any questions, feel free to shoot. Thanks!

**Comment by hanabi5** [comment]
No longer Work in progress.

Waiting feedback on:
- [ ] Gnomeregan doors keep their state after crash/restart
- [ ] Gnomeregan event boss, **Grubbis**, after kill and crash/restart, is saved and the rumble gameobjects keep their previous state.
- [ ] Stratholme traps and doors keep their state saved after a restart. Particularly the ones mention on the issue.

Although not needed to be tested, I did add the missing gameobjects from Deadmines.

**Comment by Lynethris** [comment]
For last changes:
- [X] Fix issues with gnomeregan and stratholmes

> > The save seems bind to the lead of the group. So when he leave the group, the DB is cleared. I observed that if the rest of the group interact with a gameobject, a new instance is saved. But after having done a restart in deadmine (after activating the last door), at the relog the door was closed.
> 
> The save isn't necessarily bound to the leader of the group. We don't account for players saved on this PR, only for instance IDs. When a player leaves a dungeon group, his instance ID is cleared. So even if other players are inside, they are still in the instance ID that the leader had, only that his ID was cleared due to him abandoning the group which is intended otherwise we would have un-wanted behaviours. Being a leader means nothing here, It's just a coincidence of your test case.
> 

I have redone some tests. You have right sorry. It's not bind to the lead. But there are always a problem with groups. When someone leave the group, the instance is remove in the DB. 

Here after that someone leave the group:

Here after a restart:

Test done:
- Setup a group with 3 players. Go to deadmines.
- Check the DB
- Open the first and 2nd door.
- One player leave the group
- Check the DB
- Do a restart
- Check the DB again

> Now, you mentioned the door being closed, but to whom was it closed, the player who left the group or the remaining players on party?

This explain why I have found the door close after my restart.

**Comment by hanabi5** [comment]
> This explain why I have found the door close after my restart.

You're right, I'll have to prevent this from happening.
I'll give news soon.

**Comment by hanabi5** [comment]
@Lynethris Hello!
I have made a change to prevent the instance saved data do be deleted when a player leaves group. Let me know what the results are. Cheers! 👍

File: src/server/game/Groups/Group.cpp [hunk 2]
```diff
@@ -753,6 +755,11 @@ void Group::Disband(bool hideDestroy /* = false */)
 
         player = ObjectAccessor::FindConnectedPlayer(citr->guid);
 
+        if (player != nullptr && !instanceId && !isBGGroup && !isBFGroup)
```
Comment (Kitzunu) [inline src/server/game/Groups/Group.cpp:758]:
```suggestion
        if (player && !instanceId && !isBGGroup && !isBFGroup)
```

**Comment by Nyeriah** [comment]
@Lynethris any issues remaining here?

**Comment by Lynethris** [comment]
> @Lynethris any issues remaining here?

I need to test last changes with groups. Except this point all work fine.

**Comment by Lynethris** [comment]
All is fine now. Tests done:

- [X] Base test: 3 char in deadmines, then crash:restart --> GO State saved
- [X] 2nd  test: 3 char, one leave, then crash/restart --> GO State stay saved
- [X] 3rd test: 3 char, group dissolve in instance --> GO save instance is clear

File: data/sql/updates/pending_db_characters/rev_1643629468629316100.sql [hunk 3]
```diff
@@ -0,0 +1,9 @@
+INSERT INTO `version_db_characters` (`sql_rev`) VALUES ('1643629468629316100');
```
Comment (Nyeriah) [inline data/sql/updates/pending_db_characters/rev_1643629468629316100.sql:1]:
```suggestion
```

File: data/sql/updates/pending_db_world/rev_1649359139539727000.sql [hunk 4]
```diff
@@ -0,0 +1,5 @@
+INSERT INTO `version_db_world` (`sql_rev`) VALUES ('1649359139539727000');
```
Comment (Nyeriah) [inline data/sql/updates/pending_db_world/rev_1649359139539727000.sql:1]:
```suggestion
```
